### PR TITLE
download all action script

### PR DESCRIPTION
**Pull request by Nguyenpd-lts**
Download all action script.

### CODE REVIEW

**Comment by okamura-asnet** [comment]
@Nguyenpd-lts 
Thank you for making the features!

I tested in my local.
Please check below.

1) About download target
I checked `src/commands/actions/scripts/downloads/all.ts`.
It looks like getting ALL Workspaces that user can use. But in this command, it would be great if we could specify single Project(Application).

The command specifications are as follows.
```
hx actions:scripts:downloads:all [PROJECT_ID]
```
- `PROJECT_ID` is an optional parameter.
- If `PROJECT_ID` isn't specified, could you make it to select `PROJECT_ID`?
- I think it will be helpful for implementing to select `PROJECT_ID`. -> `hx datastores:get [PROJECT_ID]` ([src/commands/datastores/get.ts](https://github.com/b-eee/hexabase-cli/blob/develop/src/commands/datastores/get.ts))

2) About action
Action is defined in Field settings. (not in Item)
I don't think it's necessary to loop Items.
We may use API below.
- Get status list
GET /api/v0/datastores/:datastore-id/statuses
https://github.com/b-eee/hexabase-cli/blob/60e71f944389ac53bfa1cb50f3627c50c2245436/src/commands/actions/get.ts#L85
- Get action list
GET /api/v0/datastores/:datastore-id/actions
URL Params：status_id
https://github.com/b-eee/hexabase-cli/blob/60e71f944389ac53bfa1cb50f3627c50c2245436/src/commands/actions/get.ts#L98

**Comment by Hokutosei** [comment]
@okamura-asnet thank you!, I already have discussion to LTS @Nguyenpd-lts , and now making actions.

File: src/commands/actions/scripts/downloads/all.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                const {data: dtItems} = await this.hexaapi.post<any>(
+                url, {
+                  "page":1,
+                  "per_page":0,
+                });
+
+                // download each script in 
+                try {
+                  const {items} = dtItems;
+                  for(let item of items){
```
Comment (iwask) [inline src/commands/actions/scripts/downloads/all.ts:79]: @Nguyenpd-lts i think we dont need this Item loop. 
actions only related to datastores or statuese in a datastore. 

datastore -> actions ( = Datastore Actions )
datastore -> statuses -> actions ( = Status Actions )

you can refer here
https://github.com/b-eee/hexabase-cli/blob/develop/src/commands/actions/get.ts#L98

cc: @Hokutosei @okamura-asnet

**Comment by Nguyenpd-lts** [comment]
I just update function download actionscript:
- I do download with this procedure:
   + I get datastore-> get action->download with type.
     ```
      - Have argument projectId or select projectId in currentworkspace
      - Get datastore: api/v0/applications/:project-id/datastores
      - Get action: /api/v0/datastores/:datastore-id/actions
      - Download /api/v0/actions/:action-id/actionscripts/download?script_type=[post or pre]
```
   + I use "display-id" as a directory name. but it doesn't exist "display-id" I use id for the directory name.

**Review by iwask — COMMENTED** [review]
@Nguyenpd-lts thank you for this PR. 

i merged other PR and fixed to adopt some func names changes, for `hexaAPI` and `hexaConfig`. 
and added cmd aliases. 
https://github.com/b-eee/hexabase-cli/pull/14/files#diff-af98d0d08d608e0d3b4cc7c332a97f75c2c3ea39ab0b6f5ea78b45203c4556fbR38

and I test this cmd, getting js files looks good, but if the action does not have ActionScript, got ERRORs from apicore. 

could you check?  and could you try to fix in apicore? 
(If you could, you should run this in your local apicore. )

this is outputs w/ this cmd. i used alias toward my local test project. 
```
$ hx asall
? Specify the output of the actionscriptdownloadtestload
√ Specify the output of the actionscript file · downloadtest
√ Select your project · 611be3b251eca76ebec45deb
√ Select your project · 611be3b251eca76ebec45deb
Error download with action:CopyItem-post
Error download with action:CopyItem-pre
Download success CreateItem
Error download with action:CreateItem-pre
Error download with action:DeleteItem-post
Error download with action:DeleteItem-pre
Error download with action:test-post
Error download with action:test-pre
Download success UpdateItem
Download success UpdateItem
Error download with action:CopyItem-post
Error download with action:CopyItem-pre
Download success CreateItem
Download success CreateItem
Error download with action:DeleteItem-post
Error download with action:DeleteItem-pre
Error download with action:UpdateItem-post
Error download with action:UpdateItem-pre
Error download with action:CopyItem-post
Error download with action:CopyItem-pre
Error download with action:CreateItem-post
Error download with action:CreateItem-pre
Error download with action:DeleteItem-post
Error download with action:DeleteItem-pre
Error download with action:UpdateItem-post
Error download with action:UpdateItem-pre
Error download with action:CopyItem-post
Error download with action:CopyItem-pre
Error download with action:CreateItem-post
Error download with action:CreateItem-pre
Error download with action:DeleteItem-post
Error download with action:DeleteItem-pre
Error download with action:UpdateItem-post
Error download with action:UpdateItem-pre
Error download with action:CopyItem-post
Error download with action:CopyItem-pre
Error download with action:CreateItem-post
Error download with action:CreateItem-pre
Error download with action:DeleteItem-post
Error download with action:DeleteItem-pre
You can specify the output folder with the '--output' flag in the future
```

javaScript files could get downloaded successfully, if it has a ActionScript, 
but if it does not exist, looks get ERROR in apicore. 

 I found ERROR logs in apicore. 
this is a part of ERROR log.  return 500. 
```
[2021-11-06T14:31:50.9686953Z] [INFO] GetActionScript took: 18.519ms
[2021-11-06T14:31:50.9694712Z] [INFO] Writing js file for a_id: 61584b5bd698d2954dae945c, tempfile: C:\Users\hiwas\AppData\Local\Temp\as-734065195.js
[GIN] 2021/11/06 - 14:31:50 | 500 |     88.8152ms |             ::1 | GET      "/download_action_script?a_id=61584b5bd698d2954dae945c&script_type=pre"
```

and as the same, linke-api returns ERROR 500
```
[GIN] 2021/11/06 - 14:31:50 | 500 |    134.8194ms |       127.0.0.1 | GET      "/api/v0/actions/61584b5bd698d2954dae945c/actionscripts/download?script_type=post"
```

apicore should return 404 if it doesn't have scripts, and CLI can ignore 404.

**Review by iwask — CHANGES_REQUESTED** [review]
pls check my comment above.
https://github.com/b-eee/hexabase-cli/pull/14#pullrequestreview-799451380

**Comment by Hokutosei** [comment]
@Nguyenpd-lts sorry to bother you. about the api `/download_action_script` responding `500`. if possible, to update the API, if no script return `404` status code instead of `500`. 

once updated the `500` to `404`, handle/ignore the status code from cli. currently, can't think of anything than to skip/ignore. since there is no actionscript to download.

**Review by iwask — APPROVED** [review]
ok, LTGM. will merge, tks!

**Comment by iwask** [comment]
i will fix apicore, should return 404 instead of 500

**Pull request merged**